src/test/java/com/fasterxml/jackson/databind/deser/TestCustomSubTypeValidator.java

```diff
@@ -0,0 +1,277 @@
+    /* Helper classes (validators)
+    /**********************************************************
+     */
+
+    static class DenyAllValidator
+        extends SubTypeValidator
+    {
+        @Override
+        public void validateSubType(DeserializationContext ctxt, JavaType type, BeanDescription beanDesc)
+            throws JsonMappingException 
+        {
+            ctxt.reportBadTypeDefinition(
+                    beanDesc,
+                    "Illegal type (%s) to deserialize: this validator does not allow deserialization at all!",
+                    type.getRawClass().getName());
+        }
+    }
+
+    static class FriendlyAnimalValidator
+        extends SubTypeValidator
+    {
+        private static Class<?>[] FRIENDLY_ANIMAL_WHITELIST = new Class<?>[] {
+            Animal.class,
+            Goat.class,
+            Sheep.class,
+            Penguin.class
+        };
+
+        @Override
+        public void validateSubType(DeserializationContext ctxt, JavaType type, BeanDescription beanDesc)
+            throws JsonMappingException 
+        {
+            super.validateSubType(ctxt, type, beanDesc);
+
+            if (type.isTypeOrSubTypeOf(Animal.class)) {
+                for (final Class<?> friendlyAnimalSpecies : FRIENDLY_ANIMAL_WHITELIST) {
+                    if (type.hasRawClass(friendlyAnimalSpecies)) {
+                        return;
+                    }
+                }
+
+                ctxt.reportBadTypeDefinition(
+                        beanDesc,
+                        "Illegal type (%s) to deserialize: prevented because this animal species is not known to be friendly",
+                        type.getRawClass().getName());
+            }
+        }
+    }
+
+    /*
+    /**********************************************************
+    /* Factory methods
+    /**********************************************************
+     */
+
+    protected ObjectMapper objectMapper(SubTypeValidator validator)
+    {
+        final DeserializerFactoryConfig config =
+                new DeserializerFactoryConfig()
+                    .withSubTypeValidator(validator);
+
+        final DeserializerFactory factory =
+                new BeanDeserializerFactory(config);
+
+        final JsonMapper mapper =
+                JsonMapper
+                    .builder()
+                    .deserializerFactory(factory)
+                    .build();
+
+        return mapper;
+    }
+
+    /*
+    /**********************************************************
+    /* Unit tests
+    /**********************************************************
+     */
+
+    public void testDefaultValidator() throws Exception
+    {
+        final ObjectMapper mapper = objectMapper();
+
+        mapper.readValue(STRING_VALUE_JSON, String.class);
+        mapper.readValue(ZOO_JSON, Zoo.class);
+        mapper.readValue(PETTING_ZOO_JSON, Zoo.class);
+    }
+
+    public void testDenyAllValidator() throws Exception
```

Comment:
Each test in this class currently test 3 different JSON inputs. Should I rather split each test into 3 separate ones?